extmod/uasyncio/core.py

```diff
@@ -0,0 +1,500 @@
+        while True:
+            try:
+                yield _io_queue.queue_read(s)
+            except CancelledError:
+                # Shutdown server
+                s.close()
+                return
+            s2, addr = s.accept()
+            s2.setblocking(False)
+            s2s = Stream(s2, {'peername': addr})
+            create_task(cb(s2s, s2s))
+
+# Helper function to start a TCP stream server, running as a new task
+# TODO could use an accept-callback on socket read activity instead of creating a task
+async def start_server(cb, host, port, backlog=5):
+    s = Server()
+    create_task(s._serve(cb, host, port, backlog))
+    return s
+
+################################################################################
+# Main run loop
+
+# Queue of Task instances
+_queue = Queue()
+
+# Task queue and poller for stream IO
+_io_queue = IOQueue()
+
+# Keep scheduling tasks until there are none left to schedule
+def run_until_complete(main_task=None):
+    global cur_task
+    excs_all = (CancelledError, Exception) # To prevent heap allocation in loop
+    excs_stop = (CancelledError, StopIteration) # To prevent heap allocation in loop
+    while True:
+        # Wait until the head of _queue is ready to run
+        dt = 1
+        while dt > 0:
+            dt = -1
+            t = _queue.peek()
+            if t:
+                # A task waiting on _queue; "ph_key" is time to schedule task at
+                dt = max(0, ticks_diff(t.ph_key, ticks()))
+            elif not _io_queue.map:
+                # No tasks can be woken so finished running
+                return
+            #print('(poll {})'.format(dt), len(_io_queue.map))
+            _io_queue.wait_io_event(dt)
+
+        # Get next task to run and continue it
+        t = _queue.pop_head()
+        cur_task = t
+        try:
+            # Continue running the coroutine, it's responsible for rescheduling itself
+            if not t.data:
+                t.coro.send(None)
+            else:
+                t.coro.throw(t.data)
+        except excs_all as er:
+            # This task is done, check if it's the main task and then loop should stop
+            if t is main_task:
+                if isinstance(er, StopIteration):
+                    return er.value
+                raise er
+            # Save return value of coro to pass up to caller
+            t.data = er
+            # Schedule any other tasks waiting on the completion of this task
+            waiting = False
+            if hasattr(t, 'waiting'):
+                while t.waiting.peek():
+                    _queue.push_head(t.waiting.pop_head())
+                    waiting = True
+                t.waiting = None # Free waiting queue head
+            # Indicate task is done
+            t.coro = None
+            # Print out exception for detached tasks
+            if not waiting and not isinstance(er, excs_stop):
+                print('task raised exception:', t.coro)
```

Comment:
`t.coro` has been set to None three lines above so log message isn't very helpful